### ISSUE TITLE
[no-unused-modules] Improve rule for 'import * as xxx from './xxx'

### ISSUE THREAD

**Issue by barbogast** [benmosher/eslint-plugin-import#1775]
If a file with named exports is imported as `import * as xxx from './xxx'` the rule assumes that all exports of this file have been used.

Would it be possible to report exports that never have been accessed?

Example:

given file-f:
```js
import * as fileG from './file-g'

const x = fileG.a
fileG.b()
```

and file-g
```js
export const a = 7 // will not be reported
export const b = () => {} // will not be reported
export const c = 5 // will be reported
```

Of course this can only work as long as a module is not access like this:
```js
import * as fileG from './fileG'
fileG['b']()
```
In this case the current behaviour would apply and nothing would be reported.

We heavily rely on the `import * as __ from __` syntax, so we would rely profit from this change. In any case, thanks a lot for all the effort that went into that plugin already!

**Comment by ljharb** [comment]
While I'd strongly suggest not relying on that syntax (for example, TS's module system is broken unless you enable synthetic imports and esModuleInterop, and once enabled, TS projects basically never need `import * as` ever again), if someone could increase the static analysis abilities of that rule, that'd be great.

Note that if you wanted to cause *more* warnings (by warning on anything that's not statically knowable), it'd need to be an option, not on by default.

**Comment by yamov** [comment]
@ljharb thanks a lot for your response. Could you please elaborate a little more (or redirect to some other source if it's already discussed there) about why you don't recommend using the `import * as` syntax. I don't fully follow the reasoning in your previous comment.

**Comment by ljharb** [comment]
@yamov `import * as ns from 'path'` means that `ns` is a frozen Module Namespace object. In JS, `import * as` is only useful for two purposes: laziness (not wanting to explicitly list all the imports you need), or metaprogramming.

**Comment by yamov** [comment]
@ljharb I see, thanks. But what are the downsides of `ns` being a frozen Module Namespace object?
Also "laziness" is probably just a matter of viewpoint :). I believe some may call it "convenience" and point out that in some cases it can be more practical. But this is apparently subjective. I would still like to understand better the underlaying technical downsides of this approach.

**Comment by ljharb** [comment]
There are no downsides; it's what the spec requires, and it's part of what TypeScript violates when you don't have esModuleInterop and synthetic imports enabled.

Explicit > implicit; always prefer explicitness and easier reading over implicitness and easier writing.

**Comment by yamov** [comment]
I see. Thanks a lot for your answers.
